pageserver/src/layered_repository/timeline.rs

```diff
@@ -2036,8 +2037,9 @@ impl Timeline {
         let _enter = info_span!("garbage collection", timeline = %self.timeline_id, tenant = %self.tenant_id, cutoff = %new_gc_cutoff).entered();
 
         // We need to ensure that no one branches at a point before latest_gc_cutoff_lsn.
-        // See branch_timeline() for details.
-        *self.latest_gc_cutoff_lsn.write().unwrap() = new_gc_cutoff;
+        // See branch_timeline() for details. This will block until the old value is
+        // no longer in use.
+        self.latest_gc_cutoff_lsn.store(new_gc_cutoff);
```

Comment:
I'd sleep better with an assert that the cutoff only increases. I don't have confidence that it won't happen, and it  will not be an easy thing to debug.

Reply:
Agreed. It was a bit harder than I thought to implement. My first thought was to return the old value from `Rcu::write`, but on return, you've already written the new value and it might already have been seen by new readers. So that's not very nice. I ended up splitting the `write` into two steps. The `Rcu::write` call now just returns a write guard object that you can use to read the current value, and the actual update and waiting for readers to finish is done in `RcuWriteGuard::store` function.